pkg/adapter/v2/main.go

```diff
@@ -56,6 +73,81 @@ type AdapterConstructor func(ctx context.Context, env EnvConfigAccessor, client
 // the multi-tenant receive adapter state
 type ControllerConstructor func(ctx context.Context, adapter Adapter) *controller.Impl
 
+type AdapterDynamicConfigOption func(*AdapterDynamicConfig)
+
+// metricDomainDefaulter is an AdapterDynamicconfig option that
+// retrieves the adater's metric domain by looking at the
+// environments metrics.DomainEnv variable first,
+// then to the component's hardcoded default metrics domain,
+// and if not present to the eventing default metrics domain.
+func metricDomainDefaulter(adc *AdapterDynamicConfig) {
+	if domain := os.Getenv(metrics.DomainEnv); domain != "" {
+		adc.metricsDomain = domain
+	} else if adc.metricsDomain == "" {
+		adc.metricsDomain = defaultMetricsDomain
+	}
+}
+
+// WithLoggerConfigMapName overrides the default logger ConfigMap name.
+func WithLoggerConfigMapName(name string) AdapterDynamicConfigOption {
+	return func(adc *AdapterDynamicConfig) {
+		adc.loggingConfigName = name
+	}
+}
+
+// WithLoggerConfigMapName overrides the default observability ConfigMap name.
+func WithObservabilityConfigMapName(name string) AdapterDynamicConfigOption {
+	return func(adc *AdapterDynamicConfig) {
+		adc.observabilityConfigName = name
+	}
+}
+
+// WithLoggerConfigMapName overrides the default tracing ConfigMap name.
+func WithTracingConfigMapName(name string) AdapterDynamicConfigOption {
+	return func(adc *AdapterDynamicConfig) {
+		adc.tracingConfigName = name
+	}
+}
+
+// NewAdapterDynamicConfig creates a dynamic configurator for adapters.
+func NewAdapterDynamicConfig(opts ...AdapterDynamicConfigOption) *AdapterDynamicConfig {
+	adc := &AdapterDynamicConfig{
+		loggingConfigName:       logging.ConfigMapName(),
+		observabilityConfigName: metrics.ConfigMapName(),
+		tracingConfigName:       tracingconfig.ConfigName,
+	}
+
+	opts = append(opts, metricDomainDefaulter)
+
+	for _, opt := range opts {
+		opt(adc)
+	}
+
+	return adc
+}
+
+// AdapterDynamicConfig keeps the ConfigMap based dynamic
+// configuration parameters for the adapter.
+type AdapterDynamicConfig struct {
+	// ConfigMap names
+	loggingConfigName       string
+	observabilityConfigName string
+	tracingConfigName       string
+
+	// Metrics domain
+	metricsDomain string
+}
+
+// AdapterConfigurator exposes methods for configuring
+// the adapter.
+type AdapterConfigurator interface {
+	CreateLogger(ctx context.Context) *zap.SugaredLogger
+	SetupMetricsExporter(ctx context.Context)
+	CreateProfilingServer(ctx context.Context) *http.Server
+	CreateCloudEventsEventStatusReporter(ctx context.Context) *crstatusevent.CRStatusEventClient
+	SetupTracing(ctx context.Context, instanceName string)
+}
```

Comment:
I think that is a good idea. What about 3 interfaces for

- `Observability`: for Logger, Metrics and Tracing.
- `Profiling`: for Profiling.
- `CloudEvents`: for CE Status reporter.

Reply:
The adapter's main is more verbose now, and although it could be simplified by grouping all customization options into a single function, I think for now it is good to have it as is for documentation/example purposes:

https://github.com/knative/eventing/blob/c94baf16b59cdd5173bdc3cbd79ce3b05bb21555/cmd/mtping/main.go#L43-L52